datafusion/core/src/datasource/file_format/mod.rs

```diff
@@ -87,6 +98,277 @@ pub trait FileFormat: Send + Sync + fmt::Debug {
         conf: FileScanConfig,
         filters: Option<&Arc<dyn PhysicalExpr>>,
     ) -> Result<Arc<dyn ExecutionPlan>>;
+
+    /// Take a list of files and the configuration to convert it to the
+    /// appropriate writer executor according to this file format.
+    async fn create_writer_physical_plan(
+        &self,
+        _input: Arc<dyn ExecutionPlan>,
+        _state: &SessionState,
+        _conf: FileSinkConfig,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        let msg = "Writer not implemented for this format".to_owned();
+        Err(DataFusionError::NotImplemented(msg))
+    }
+}
+
+/// `AsyncPutWriter` is an object that facilitates asynchronous writing to object stores.
+/// It is specifically designed for the `object_store` crate's `put` method and sends
+/// whole bytes at once when the buffer is flushed.
+pub struct AsyncPutWriter {
```

Comment:
@tustvold  and/or @crepererum  (my go to people for rust async / stream expertise) I wonder if you have some time to review this code that adapts DataFusion to use the object_store put multi part features to do streaming writes

Pretty exciting!

Reply:
I'm confused to see this here, we already provide AsynWrite from ObjectStore, so I'm not sure why we are re-implementing the buffering here? Am i missing something?

Reply:
We hypothesize that the consistent use of `put_multipart` for every put operation might adversely impact the cloud side, as it anticipates files exceeding a specific size (for example, 5MB for AWS). To mitigate this, we've developed a wrapper for the `put` operation that standardizes the write operation on `AsyncWrite`.

Reply:
Oh yes 100%, my suggestion was be to type-erase at the level of the BatchSerializer and then have different impls for the different write modes. The async Read + abort interface feels a tad over complicated, at least imo.